velox/type/Type.cpp

```diff
@@ -611,33 +615,48 @@ bool Type::containsUnknown() const {
 
 namespace {
 
-using CustomTypeFactory =
-    std::function<TypePtr(std::vector<TypePtr> childTypes)>;
-
-std::unordered_map<std::string, CustomTypeFactory>& typeFactories() {
-  static std::unordered_map<std::string, CustomTypeFactory> factories;
+std::unordered_map<std::string, std::unique_ptr<const CustomTypeFactories>>&
+typeFactories() {
+  static std::
+      unordered_map<std::string, std::unique_ptr<const CustomTypeFactories>>
+          factories;
   return factories;
 }
 } // namespace
 
-void registerType(const std::string& name, CustomTypeFactory factory) {
+void registerType(
+    const std::string& name,
+    std::unique_ptr<const CustomTypeFactories> factories) {
   auto uppercaseName = boost::algorithm::to_upper_copy(name);
-  typeFactories().emplace(uppercaseName, factory);
+  typeFactories().emplace(uppercaseName, std::move(factories));
 }
 
 bool typeExists(const std::string& name) {
   auto uppercaseName = boost::algorithm::to_upper_copy(name);
   return typeFactories().count(uppercaseName) > 0;
 }
 
-TypePtr getType(const std::string& name, std::vector<TypePtr> childTypes) {
+auto getTypeEntry(const std::string& name) {
```

Comment:
I find it confusing to use auto as return type of a function. It is hard to understand what is the actual type. Any chance you could spell out the type?

Reply:
Replaced `auto` with `const CustomTypeFactories*`. Thank you for the suggestion!